compiler/compiler/src/run_with_ledger.rs

```diff
@@ -0,0 +1,299 @@
+// Copyright (C) 2019-2025 Provable Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use leo_errors::{BufferEmitter, ErrBuffer, Handler, LeoError, Result, WarningBuffer};
+
+use aleo_std_storage::StorageMode;
+use anyhow::anyhow;
+use snarkvm::{
+    prelude::{
+        Address,
+        Execution,
+        Ledger,
+        PrivateKey,
+        Process,
+        ProgramID,
+        TestnetV0,
+        Transaction,
+        VM,
+        Value,
+        store::{ConsensusStore, helpers::memory::ConsensusMemory},
+    },
+    synthesizer::program::ProgramCore,
+};
+
+use rand_chacha::{ChaCha20Rng, rand_core::SeedableRng as _};
+use serde_json;
+use std::{fmt, str::FromStr as _};
+
+type CurrentNetwork = TestnetV0;
+
+/// Programs and configuration to run.
+pub struct Config {
+    pub seed: u64,
+    pub min_height: u32,
+    pub programs: Vec<Program>,
+}
+
+/// A program to deploy to the ledger.
+#[derive(Clone, Debug, Default)]
+pub struct Program {
+    pub bytecode: String,
+    pub name: String,
+}
+
+/// A particular case to run.
+#[derive(Clone, Debug, Default)]
+pub struct Case {
+    pub program_name: String,
+    pub function: String,
+    pub private_key: Option<String>,
+    pub input: Vec<String>,
+}
+
+/// The status of a case that was run.
+#[derive(Clone, PartialEq, Eq)]
+pub enum Status {
+    None,
+    Aborted,
+    Accepted,
+    Rejected,
+    Halted(String),
+}
+
+impl fmt::Display for Status {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Status::Halted(s) => write!(f, "halted ({s})"),
+            Status::None => "none".fmt(f),
+            Status::Aborted => "aborted".fmt(f),
+            Status::Accepted => "accepted".fmt(f),
+            Status::Rejected => "rejected".fmt(f),
+        }
+    }
+}
+
+/// All details about the result of a case that was run.
+pub struct CaseOutcome {
+    pub status: Status,
+    pub verified: bool,
+    pub errors: ErrBuffer,
+    pub warnings: WarningBuffer,
+    pub execution: String,
+}
+
+/// Run the functions indicated by `cases` from the programs in `config`.
+// Currently this is used both by the test runner in `test_execution.rs`
```

Comment:
Maybe one day our execution tests are entirely in the test framework and then we can move it out :)